### PR TITLE
[pkg] detect bad packages while interning

### PR DESCRIPTION
**Pull request by PaliC**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #73456
* #71731
* __->__ #70955

It's possible to call intern on packages which don't follow normal python conventions (https://docs.python.org/3/reference/import.html#packages). We do not support interning these types of packages, however, debugging errors caused by them can be a pain. This pr throws an error for these incorrectly written packages to make them easily detectable.

Differential Revision: [D33466521](https://our.internmc.facebook.com/intern/diff/D33466521)

### CODE REVIEW

**Comment by PaliC** [comment]
@PaliC has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33466521).

**Comment by PaliC** [comment]
@PaliC has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33466521).

**Comment by PaliC** [comment]
@PaliC has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33466521).

File: torch/package/package_exporter.py [hunk 1]
```diff
@@ -493,6 +499,29 @@ def save_module(self, module_name: str, dependencies=True):
...
         self._intern_module(module_name, dependencies)
 
+    def _is_package(
+        self,
+        module_obj: types.ModuleType
+    ):
+        """Checks if module_obj is a package and throws an error, if module_obj is an
+            invalid package.
+        """
+        error = BadPackageError(f"{module_obj.__name__} is an improperly formed package "
```
Comment (suo) [inline torch/package/package_exporter.py:509]: Can we improve the error message to explain why the package is improperly formed? Most of our users probably don't know anything about the Python import/module system.

File: torch/package/package_exporter.py [hunk 2]
```diff
@@ -493,6 +499,29 @@ def save_module(self, module_name: str, dependencies=True):
...
+                                f",so it cannot be interned. Please mock or extern {module_obj.__name__}.")
+        if not hasattr(module_obj, "__path__"):
+            return False
+        elif isinstance(module_obj.__path__, list):  # type: ignore [attr-defined]
+            return True
+        try:
+            for i in module_obj.__name__:
```
Comment (suo) [inline torch/package/package_exporter.py:516]: I don't understand what we're testing here, I assume it has something to do with iterableness; can you add a comment explaining

**Review by suo — CHANGES_REQUESTED** [review]

**Comment by PaliC** [comment]
@PaliC has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33466521).

**Review by suo — APPROVED** [review]
lgtm!

**Comment by PaliC** [comment]
@PaliC has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33466521).

**Comment by PaliC** [comment]
@PaliC has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33466521).

**Comment by PaliC** [comment]
@PaliC has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33466521).

**Pull request closed without merging**